### DISCUSSION TITLE
How to 'build-from-source' from a specific checkout, not HEAD?

### DISCUSSION THREAD

**Discussion by edrozenberg** [Homebrew/discussions#1230] [Tap maintenance and Homebrew development]
I'm trying to git bisect midnight-commander (`mc`). The `4.8.25` release works fine on Mac, `4.8.26` is partially broken (does not launch in any dir other than homedir). Want to figure out which commit broke it so I have some chance of getting the mc project to look at it.

I tried but cannot figure out any way to `brew install --build-from-source` from a specific git tag/checkout, and not `--HEAD`.

For ex. I'd want to force `brew` to build from a directory I create using `git clone --depth 1 <repo_url>; cd <git dir>; git reset --hard <COMMIT-SHA-ID>`

It seems currently there's no way to tell `brew` to just build the source from where it tell it, instead of having it do its own HEAD checkout.

**Comment by SMillerDev** [comment]
You can specify a git url + revision as the URL and rebuild it normally. Normally we also use a tag in that structure but it's not required.

**Reply by edrozenberg** [reply]
That sounds exactly like what I need but I haven't figured out the right way to spec it, or it doesn't work.

Tried editing `midnight-commander.rb` as follows:

```
  head do
    url "https://github.com/MidnightCommander/mc/commit/4df5bbf9ef9634b00f3c19ad55425f3a2142463f"
```

```
==> Downloading https://github.com/MidnightCommander/mc/commit/4df5bbf9ef9634b00
##O#- #                                                                       
Warning: Cannot verify integrity of 'a962abdddaadbe705b23a8961d2b5167b476d07f8f4db060d15da44f482e8ff0--4df5bbf9ef9634b00f3c19ad55425f3a2142463f'.
No checksum was provided for this resource.
For your reference, the checksum is:
  sha256 "5c95cbcf7ffe39a912d5114a7d6b8fcc17063a797fa4a1e2048b2567000c8cd1"
==> ./autogen.sh
Last 15 lines from /Users/eduardr/Library/Logs/Homebrew/midnight-commander/01.autogen.sh:
2021-04-09 08:35:53 -0700

./autogen.sh


READ THIS: https://docs.brew.sh/Troubleshooting
```

**Reply by SMillerDev** [reply]
You need to change the default URL, see for example:
https://github.com/Homebrew/homebrew-core/blob/master/Formula/acmetool.rb#L4-L6

**Reply by edrozenberg** [reply]
@SMillerDev thanks, tried but failed. I'll give up on this approach for now. Maybe "someday" I'll send a PR for a built-in brew feature that makes it easy to build from source a specific tag/release/commit.

For reference tried modifying as below:
```
class MidnightCommander < Formula
  desc "Terminal-based visual file manager"
  homepage "https://www.midnight-commander.org/"
  url "https://github.com/MidnightCommander/mc.git",
      revision: "171957980e17331d6b7dab6a343db9e1f3fef49c"
  license "GPL-3.0-or-later"
```
`brew install --build-from-source mc`
`Error: invalid attribute for formula 'midnight-commander': version (nil)`

**Reply by Rylan12** [reply]
Since there's no `tag`, Homebrew doesn't know how to automatically extract the version so you'll need to specify it yourself with a `version` spec. Since you're just doing some testing it probably doesn't matter much what you set it to.

```ruby
class MidnightCommander < Formula
  desc "Terminal-based visual file manager"
  homepage "https://www.midnight-commander.org/"
  url "https://github.com/MidnightCommander/mc.git",
      revision: "171957980e17331d6b7dab6a343db9e1f3fef49c"
  # I chose "4.8.26" because that's the latest stable version but you
  # could also choose e.g. "1.0.0" or whatever you want
  version "4.8.26"
  license "GPL-3.0-or-later"
```

**Reply by edrozenberg** [reply]
@Rylan12 thanks, very helpful! Sounds like the version is whatever you want it to be so I guess I could set it to something like `20200123_1234567` (matching the date/hash of the commit I'm using). 

Meanwhile I took the manual build approach & w/git bisect and found the "bad" commit - related to the `mc` subshell. My short term workaround is `mc --nosubshell` to avoid whatever the subshell issue is, and I'll need to submit the issue to the mc project and see if they'll take it up.

**Reply by Rylan12** [reply]
> @Rylan12 thanks, very helpful! Sounds like the version is whatever you want it to be so I guess I could set it to something like `20200123_1234567` (matching the date/hash of the commit I'm using).

Yeah, it matters more for when you're creating an actual formula to be distributed (in which case you should be using a `tag` not just a `revision` so you won't have the issue) but if you're just testing a formula out it can be whatever. You have no idea how many test formulae called "foo" I've made... 😅

**Comment by carlocab** [comment]
This seems like a really painful way of doing it, as there are a few hundred commits in between release tags. I suggest building outside of `brew` so you can at least make use of `git bisect`.

**Reply by edrozenberg** [reply]
@carlocab thanks, probably the right way to go. I'll have to "reverse engineer" what the brew `mc` formula does for the source build (config params, an ENV tweak, etc). Lot of experience building stuff on Linux, not much on Mac. I'll figure it out eventually.

**Reply by carlocab** [reply]
The build flags you can see in the formula. The only other important thing that `brew` does silently should be equivalent to
```sh
export PKG_CONFIG_PATH="$(brew --prefix openssl@1.1)/lib/pkgconfig"
```

It would also be good to know if this workaround is still needed:
```ruby
ENV["ac_cv_func_utimensat"] = "no" if MacOS.version >= :high_sierra
```

**Reply by edrozenberg** [reply]
@carlocab thanks again. Tried one build with and without setting `ENV["ac_cv_func_utimensat"]` - doesn't make a difference to this particular problem. Maybe it's required to fix something else - I don't know what the original intent of that tweak was.

According to `git bisect` mc was broken sometime `4.8.25` and `4.8.26` by the following commit:
```
ff0fc17ae3c88128801644f7ff5f4628881206f6 is the first bad commit
commit ff0fc17ae3c88128801644f7ff5f4628881206f6
Author: Eric Roberts <ericmrobertsdeveloper@gmail.com>
Date:   Sat Sep 19 11:58:51 2020 +0300

    Ticket #4114: implement persistent command line buffer for subshell.
```

I'll have to report it to the `mc` project and see if there's something they can do about it (their level of caring about MacOS is unknown, hopefully they care...).

Used the following build script:
```bash
#!/usr/local/bin/bash

# Set var suggested on HomeBrew discussion board
export PKG_CONFIG_PATH="$(brew --prefix openssl@1.1)/lib/pkgconfig"

# Setting used by homebrew's formula
export ac_cv_func_utimensat="no"

# Go to the source
cd ~/Developer/External/mc

# Get short commit hash
hash=`git rev-parse --short=7 HEAD`

# Clean the source
make clean

# Bootstrap the source
./autogen.sh

# Configure the source
./configure                                                  \
  --disable-dependency-tracking                              \
  --disable-silent-rules                                     \
  --prefix=/usr/local/Cellar/midnight-commander/HEAD-${hash} \
  --without-x                                                \
  --with-screen=slang                                        \
  --enable-vfs-sftp

# Compile and install
make
make install
```

And my bisect session:
```
cd ~/Developer/External/mc
$ git bisect start
$ git bisect bad                           (HEAD-9436bd1)
$ git bisect good tags/4.8.25              (HEAD-1719579)
$ git bisect bad                           (HEAD-6f96cad)
$ git bisect bad                           (HEAD-2225af8)
$ git bisect bad                           (HEAD-2dc692c)
$ git bisect bad                           (HEAD-09473d8)
$ git bisect good                          (HEAD-794f94c)
$ git bisect bad                           (HEAD-c22387d)
$ git bisect good                          (HEAD-f63be5a)
$ git bisect bad                           (HEAD-ff0fc17)
```

**Reply by edrozenberg** [reply]
Problem is subshell related, because if I launch as `mc --nosubshell` in any directory as a normal user it launches instantly (as it should), but then there's no shell available with `ctrl-o`.
